### PR TITLE
[WFLY-13267] Convert EJB3 subsystem transformers to use chained transformers

### PR DESCRIPTION
**Pull request by rachmatowicz**
This PR does the following:
- adds in transformer tests for the ejb3 subsystem for EAP versions 7.0.0, 7.1.0 and 7.2.0
- refactors the transformers to use the chained transformer approach, to simplify understanding of the transformers and make maintenance easier  

For more details, see: https://issues.redhat.com/browse/WFLY-13267

### CODE REVIEW

**Comment by rachmatowicz** [comment]
Hi @pferraro , any feedback on how to improve this would be appreciated. In other words, waiting on review.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java
Comment (bstansberry) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java:126]: This attribute is an alias.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemXMLPersister.java
Comment (bstansberry) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemXMLPersister.java:497]: AFAICT the changes to this file should not be necessary as the behavior of these AttributeDefinitions is to not write anything if the attribute isn't defined.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java
Comment (bstansberry) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java:86]: I think this move introduces merge conflicts because the whole method is gone now.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemXMLPersister.java
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemXMLPersister.java:497]: OK, i'll remove the changes. Thanks.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java:86]: Not sure I understand. When I open the file in the PR, I see the method. Do you mean it disappears after being merged?

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java:126]: This one is tricky. There is only one instance of ALIAS in the whole of Wildfly and its this one. I discussed with PF and RH and they said don't use it. If I do use it, the transformer tests fail. Is there some description somewhere of how to handle attributes with an ALIAS flag? The description above the definition says, "this attribute is an alias to something else". How are aliases handled in general?

Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java:126]: What is happening is the model comparison fails: in 6.0.0, statistics-enabled=expression and enable-statistics=expression; in 5.0.0, statistics-enabled=expression and enable-statistics=undefined. In 6.0.0, enable-statistics is flagged as ALIAS and there are custom read-attribute and write-attribute handlers for enable-statistics which instead read and write from statistics-enabled. This explains why the values are equal in 6.0.0. Exactly the same setup in 5.0.0 (EAP 7.2.0, 7.1.0). The operation transformers are creating the correct model, but the resource transformers are creating a model where enable-statistics is undefined. Is this a case where a value needs to be injected if it is undefined?

Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java:126]: I created an attribute converter for enable-statistics which will check if it is undefined and if it is, set the value to the value of statistics-enabled. I assumed that the undefined value was being created on the initial model transfer; this seems to fix it.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java
Comment (bstansberry) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java:87]: When I wrote that comment Github was reporting this PR had a merge conflict.  At the time, github was showing the change this PR was making to this file was moving this method from lower in the file to here. I assumed that was the merge conflict because Paul had done a different change to this file that removed this method altogether.

It looks like now the conflict has been resolved but the resolution resulted in restoring this method. I don't think it should be restored. Give the rest of Paul's change all it's doing here is what the superclass implementation will do.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java
Comment (bstansberry) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java:126]: You've revealed that a bug fix in WF 15 / EAP 7.3 (https://issues.redhat.com/browse/WFCORE-4183) fixes another problem besides the one it targeted. :-) When you invoke read-resource on an EAP 7.2 or earlier controller you get the value of aliases even though the op has the 'include-aliases="false" param set.

This is better to correct in DefaultModelFixer as opposed to changing the production code to send a value that the legacy host does not need. The result from the read-resource operation on the legacy controller is not correct due to that bug, so better to correct it in the test.

With https://github.com/bstansberry/wildfly/commit/fd3d27fe6cdf1f3fe453b72993026e35ac3c7c21 added to what you have here the EJB3 subsystem tests pass. Feel free to squash that in if you agree it is correct.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java:87]: Removed the method populateModel() as it is no longer in master..

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemRootResourceDefinition.java:126]: Added in your code snippet to fix the issue in the ModelFixer. Thanks, Brian.

**Review by bstansberry — APPROVED** [review]

**Comment by rachmatowicz** [comment]
I think there is a problem here with overwriting changes from WFLY-13015, fixing them now.

**Comment by rachmatowicz** [comment]
This is going to take a while; now my tests are failing because of the changes to DiscardAttributeChecker.DEFAULT replacing the original default values. I've got to go through and check each change...

**Review by chengfang — APPROVED** [review]

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java:152]: @chengfang Thanks, that instance is fixed in the current state of the PR.

**Comment by bstansberry** [comment]
I removed the "Ready for merge" based on the last couple comments.

**Comment by rachmatowicz** [comment]
It's ready now...I just missed a couple of places where Pauls previous commit got reverted by my commit during the merge

**Pull request merged**
